cli/explain/compactgraph/compactgraph.go

```diff
@@ -0,0 +1,478 @@
+package compactgraph
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"slices"
+	"sort"
+	"sync"
+
+	"github.com/buildbuddy-io/buildbuddy/proto/spawn_diff"
+	"github.com/klauspost/compress/zstd"
+	"golang.org/x/exp/maps"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/protobuf/encoding/protodelim"
+
+	spawnproto "github.com/buildbuddy-io/buildbuddy/proto/spawn"
+)
+
+// CompactGraph represents the compact execution log as a directed acyclic graph of spawns.
+// The keys are the output paths of all spawns in the graph and each spawn contains references to other execution log
+// entries, such as input files, directories, and input sets. The edges between spawns are tracked implicitly by
+// matching the output paths of the spawns to the input paths of the referenced entries.
+type CompactGraph map[string]*Spawn
```

Comment:
The graph representation treats the digests as opaque byte blobs, so it doesn't need to store the hash function name at the top level. Since we return `Digest`s as part of the diffs, I am now passing the name down to the individual `File` nodes during creation.

Since an input can be consumed by a large number of spawns, I think that we would want to avoid building such an index. When it happens, we should definitely make this a struct, but I also don't want to promise API stability for this package yet :-D